packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +50,430 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
+    "objectType",
+    "property",
+    "default",
+    "optional",
+    "indexed",
+    "mapTo",
+  ]);
+  const invalidKeysUsed = filterInvalidKeys(schema, validKeys);
+  assert(
+    !invalidKeysUsed.length,
+    `Unexpected field(s) found on the schema for property '${displayedName}': '${invalidKeysUsed.join("', '")}'.`,
+  );
 }
 
+/**
+ * Get the keys of an object that are not part of the provided valid keys.
+ */
+function filterInvalidKeys(object: Record<string, unknown>, validKeys: Set<string>): string[] {
+  return Object.keys(object).filter((key) => !validKeys.has(key));
+}
+
+/**
+ * Determine whether a property always is implicitly optional (nullable).
+ */
+function optionalIsImplicitlyTrue(type: string, objectType: string | undefined): boolean {
+  return (
+    type === "mixed" ||
+    objectType === "mixed" ||
+    type === "object" ||
+    (type === "dictionary" && isUserDefined(objectType))
+  );
+}
+
+/**
+ * Determine whether a property always is implicitly non-optional (non-nullable).
+ */
+function optionalIsImplicitlyFalse(type: string, objectType: string | undefined): boolean {
+  return (type === "list" || type === "set" || type === "linkingObjects") && isUserDefined(objectType);
+}
+
+/**
+ * Determine whether a string ends with a shorthand collection ('[]' or '{}' or '<>').
+ */
+function endsWithCollection(input: string): boolean {
+  const end = input.substring(input.length - 2);
+  return !!COLLECTION_SHORTHAND_TO_NAME[end];
+}
+
+/**
+ * Determine whether shorthand notation is being used.
+ */
+function isUsingShorthand(input: string | undefined): boolean {
+  if (!input) {
+    return false;
+  }
+  return endsWithCollection(input) || input.endsWith("?");
+}
+
+/**
+ * Get a custom error message containing the shorthands if used.
+ */
+function errMessageIfUsingShorthand(name: ObjectAndPropertyName, input: string | undefined): string {
+  const shorthands: string[] = [];
+
+  if (input && endsWithCollection(input)) {
+    shorthands.push(input.substring(input.length - 2));
+    input = input.substring(0, input.length - 2);
+  }
+
+  if (input?.endsWith("?")) {
+    shorthands.push("?");
+  }
+
+  return shorthands.length
+    ? errMessage(name, `Cannot use shorthand '${shorthands.join("' and '")}' in combination with using an object.`)
+    : "";
+}
+
+/**
+ * Get an error message for an invalid property schema.
+ */
+function errMessage(name: ObjectAndPropertyName, message: string): string {
+  return `Invalid schema for property '${name.objectName}.${name.propertyName}': ${message}`;
+}
+
+/**
+ * Throw an error caused by an invalid property schema.
+ */
+function error(name: ObjectAndPropertyName, message: string): never {
+  // TODO: Create a SchemaParseError that extends Error?
+  throw new Error(errMessage(name, message));
+}
+
+/**
+ * Extract the base type and the type argument from a generic string notation.
+ */
 export function extractGeneric(type: string): { typeBase: string; typeArgument?: string } {
```

Comment:
Is this dead code?

Reply:
Seems to be only referenced by tests. Does anyone know if this was needed for something upcoming..??